### PR TITLE
MINIFICPP-114 Consolidate JSON API use to RapidJSON

### PR DESCRIPTION
**Pull request by calebj**
Thank you for submitting a contribution to Apache NiFi - MiNiFi C++.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced
     in the commit message?

- [x] Does your PR title start with MINIFI-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [x] Is your initial contribution a single, squashed commit?

### For code changes:
- [x] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [x] If applicable, have you updated the LICENSE file?
- [x] If applicable, have you updated the NOTICE file?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

**Comment by achristianson** [comment]
@calebj whoa, nice!!! This is something I have been thinking about for a while, but was really back burner.

This is great, will check it out and let you know if I find anything.

File: extensions/http-curl/protocols/RESTProtocol.cpp [hunk 1]
```diff
@@ -32,46 +33,57 @@ namespace minifi {
...
           // set the identifier if one exists
-          if (request.isMember("operationid")) {
-            new_command.ident = request["operationid"].asString();
+          if (request.HasMember("operationid")) {
+            if (request["operationid"].IsNumber())
+              new_command.ident = std::to_string(request["operationid"].GetInt64());
+            else if (request["operationid"].IsString())
+              new_command.ident = request["operationid"].GetString();
+            else
+              throw(Exception(SITE2SITE_EXCEPTION, "Invalid type for operationid"));
```
Comment (phrocker) [inline extensions/http-curl/protocols/RESTProtocol.cpp]: This exception should not be here.

File: extensions/http-curl/protocols/RESTProtocol.cpp [hunk 2]
```diff
@@ -81,48 +93,117 @@ const C2Payload RESTProtocol::parseJsonResponse(const C2Payload &payload, const
...
-      if (child_vector.second.size() == 1) {
-        if (child_vector.second.at(0).isMember(child_vector.first)) {
-          json_payload[child_vector.first] = child_vector.second.at(0)[child_vector.first];
-        } else {
-          json_payload[child_vector.first] = child_vector.second.at(0);
-        }
-      }
-    }
-  }
+void setJsonStr(const std::string& key, const std::string& value, rapidjson::Value& parent, rapidjson::Document::AllocatorType& alloc) { // NOLINT
```
Comment (phrocker) [inline extensions/http-curl/protocols/RESTProtocol.cpp]: why is there no linter here? We want this linted.

Comment (phrocker) [inline extensions/http-curl/protocols/RESTProtocol.cpp]: Why is setJsonStr here without any scope and should it be elsewhere as a result?

File: extensions/http-curl/protocols/RESTProtocol.cpp [hunk 3]
```diff
@@ -81,48 +93,117 @@ const C2Payload RESTProtocol::parseJsonResponse(const C2Payload &payload, const
...
+  const char* c_val = value.c_str();
 
+  keyVal.SetString(c_key, key.length()), alloc;
+  valueVal.SetString(c_val, value.length(), alloc);
+
+  parent.AddMember(keyVal, valueVal, alloc);
+}
+
+rapidjson::Value getStringValue(const std::string& value, rapidjson::Document::AllocatorType& alloc) { // NOLINT
+  rapidjson::Value Val;
```
Comment (phrocker) [inline extensions/http-curl/protocols/RESTProtocol.cpp]: This function seems almost unnecessary. Why not use a macro so that it's always inlined?

**Comment by phrocker** [comment]
@calebj Thanks for the contribution, but before this can be merged can you provide us some tractable benchmarks? Are we making a library change just to make a library change based on evidence? https://github.com/google/benchmark is a great tool that will help.

File: extensions/http-curl/protocols/RESTProtocol.cpp [hunk 4]
```diff
@@ -81,48 +93,117 @@ const C2Payload RESTProtocol::parseJsonResponse(const C2Payload &payload, const
...
+    if (use_sub_option) {
+      rapidjson::Value sub_key = getStringValue(payload_content.name, alloc);
+      target.AddMember(sub_key, payload_content_values, alloc);
+    }
+  }
+}
+
+std::string RESTProtocol::serializeJsonRootPayload(const C2Payload& payload) {
+  rapidjson::Document json_payload(rapidjson::kObjectType);
+  rapidjson::Document::AllocatorType &alloc = json_payload.GetAllocator();
```
Comment (phrocker) [inline extensions/http-curl/protocols/RESTProtocol.cpp]: Can the allocator be saved for future payloads?

File: extensions/http-curl/protocols/RESTProtocol.cpp [hunk 5]
```diff
@@ -81,48 +93,117 @@ const C2Payload RESTProtocol::parseJsonResponse(const C2Payload &payload, const
...
+  mergePayloadContent(json_payload, payload, alloc);
+
+  for (const auto &nested_payload : payload.getNestedPayloads()) {
+    rapidjson::Value np_key = getStringValue(nested_payload.getLabel(), alloc);
+    rapidjson::Value np_value = serializeJsonPayload(nested_payload, alloc);
+    json_payload.AddMember(np_key, np_value, alloc);
+  }
+
+  rapidjson::StringBuffer buffer;
+  rapidjson::PrettyWriter<rapidjson::StringBuffer> writer(buffer);
```
Comment (phrocker) [inline extensions/http-curl/protocols/RESTProtocol.cpp]: payloads do not need to be "pretty written" per se.  Is this the default writer? Does this add spaces and additional bytes that increase the size of the uncompressed payload?

File: extensions/http-curl/protocols/RESTReceiver.cpp [hunk 6]
```diff
@@ -65,38 +65,8 @@ void RESTReceiver::initialize(const std::shared_ptr<core::controller::Controller
   }
 }
 int16_t RESTReceiver::heartbeat(const C2Payload &payload) {
-  std::string operation_request_str = getOperation(payload);
```
Comment (phrocker) [inline extensions/http-curl/protocols/RESTReceiver.cpp]: I assume serializeJsonRootPayload has evolved to nest the payloads? Have you tested this with a C2 server? We should do that before merging.

File: extensions/http-curl/sitetosite/PeersEntity.h [hunk 7]
```diff
@@ -35,73 +41,82 @@ class PeersEntity {
...
-            int port = 0, flowFileCount = 0;
-            bool secure = false;
+      if (!ok) {
+          std::stringstream ss;
+          ss << "Failed to parse archive lens stack from JSON string with reason: "
+             << rapidjson::GetParseError_En(ok.Code())
+             << " at offset " << ok.Offset();
+          std::string exception_msg = ss.str();
+  
+          throw Exception(ExceptionType::GENERAL_EXCEPTION, exception_msg.c_str());
```
Comment (phrocker) [inline extensions/http-curl/sitetosite/PeersEntity.h]: Thanks for doing this. I was noticing this a few days ago.

File: extensions/libarchive/UnfocusArchiveEntry.cpp [hunk 8]
```diff
@@ -67,55 +66,41 @@ void UnfocusArchiveEntry::onTrigger(core::ProcessContext *context, core::Process
...
+  {
+    ArchiveStack archiveStack;
+    {
+      std::string existingLensStack;
+    
+      if (flowFile->getAttribute("lens.archive.stack", existingLensStack)) {
+        logger_->log_info("FocusArchiveEntry loading existing lens context");
+
+        try {
+          archiveStack.loadJsonString(existingLensStack);
```
Comment (phrocker) [inline extensions/libarchive/UnfocusArchiveEntry.cpp]: Should we not give the option to transfer this to failure?

File: libminifi/src/RemoteProcessorGroupPort.cpp [hunk 9]
```diff
@@ -264,20 +264,27 @@ void RemoteProcessorGroupPort::refreshRemoteSite2SiteInfo() {
...
-      Json::Value value;
-      Json::Reader reader;
-      bool parsingSuccessful = reader.parse(controller, value);
-      if (parsingSuccessful && !value.empty()) {
-        Json::Value controllerValue = value["controller"];
-        if (!controllerValue.empty()) {
-          Json::Value port = controllerValue["remoteSiteListeningPort"];
-          if (client_type_ == sitetosite::CLIENT_TYPE::RAW && !port.empty())
-            this->site2site_port_ = port.asInt();
+
```
Comment (phrocker) [inline libminifi/src/RemoteProcessorGroupPort.cpp]: If we're going to change the JSON code we should decouple it from this class.

File: libminifi/src/core/reporting/SiteToSiteProvenanceReportingTask.cpp [hunk 10]
```diff
@@ -53,54 +56,92 @@ void SiteToSiteProvenanceReportingTask::initialize() {
   RemoteProcessorGroupPort::initialize();
 }
 
+void setJsonStr(const std::string& key, const std::string& value, rapidjson::Value& parent, rapidjson::Document::AllocatorType& alloc) { // NOLINT
```
Comment (phrocker) [inline libminifi/src/core/reporting/SiteToSiteProvenanceReportingTask.cpp]: why is this here too?

File: extensions/libarchive/FocusArchiveEntry.cpp [hunk 11]
```diff
@@ -112,77 +109,45 @@ void FocusArchiveEntry::onTrigger(core::ProcessContext *context, core::ProcessSe
...
+    ArchiveStack archiveStack;
 
     std::string existingLensStack;
 
     if (flowFile->getAttribute("lens.archive.stack", existingLensStack)) {
       logger_->log_info("FocusArchiveEntry loading existing lens context");
-      if (!reader.parse(existingLensStack, lensStack)) {
-        logger_->log_error("FocusArchiveEntry JSON parse error: %s", reader.getFormattedErrorMessages());
+      try {
+        archiveStack.loadJsonString(existingLensStack);
```
Comment (phrocker) [inline extensions/libarchive/FocusArchiveEntry.cpp]: Same here. Is this something we want to rollback or transfer to failure?

File: extensions/libarchive/FocusArchiveEntry.cpp [hunk 12]
```diff
@@ -295,7 +260,9 @@ int64_t FocusArchiveEntry::ReadCallback::process(std::shared_ptr<io::BaseStream>
   return nlen;
 }
 
-FocusArchiveEntry::ReadCallback::ReadCallback(core::Processor *processor,fileutils::FileManager *file_man, ArchiveMetadata *archiveMetadata)
```
Comment (phrocker) [inline extensions/libarchive/FocusArchiveEntry.cpp]: Our line length is 200. We haven't merged your changes to lint the extensions but you may want to limit these types of changes to reduce later changes.

File: extensions/rocksdb-repos/CMakeLists.txt [hunk 13]
```diff
@@ -23,7 +23,7 @@ cmake_minimum_required(VERSION 2.6)
 set(CMAKE_EXE_LINKER_FLAGS "-Wl,--export-all-symbols")
 set(CMAKE_SHARED_LINKER_FLAGS "-Wl,--export-symbols")
 
-include_directories(../../libminifi/include  ../../libminifi/include/core  ../../thirdparty/spdlog-20170710/include ../../thirdparty/concurrentqueue ../../thirdparty/yaml-cpp-yaml-cpp-0.5.3/include ../../thirdparty/civetweb-1.9.1/include ../../thirdparty/jsoncpp/include ../../thirdparty/rocksdb/include  ../../thirdparty/)
```
Comment (phrocker) [inline extensions/rocksdb-repos/CMakeLists.txt]: thanks for catching this. We've all been copying and pasting these so at some point we need to clean them up and I'm glad you caught this.

Comment (achristianson) [reply extensions/rocksdb-repos/CMakeLists.txt]: @phrocker 1000x we need some standard includes/functions for the cmake files.

File: extensions/http-curl/protocols/RESTProtocol.cpp (hunk 5 above)
Comment (calebj) [reply extensions/http-curl/protocols/RESTProtocol.cpp]: It used to use the formatted "pretty" writer before the change; and yes, it does. Some of the tests might have to change to work with unformatted JSON.

File: libminifi/src/core/reporting/SiteToSiteProvenanceReportingTask.cpp (hunk 10 above)
Comment (calebj) [reply libminifi/src/core/reporting/SiteToSiteProvenanceReportingTask.cpp]: It's just a shortcut wrapper for adding members from strings. I ran into a lot of issues with strings being read as garbage. I'll move them to a central location.

File: extensions/http-curl/protocols/RESTProtocol.cpp (hunk 4 above)
Comment (calebj) [reply extensions/http-curl/protocols/RESTProtocol.cpp]: I believe it's supposed to be per-document.

File: extensions/http-curl/protocols/RESTProtocol.cpp (hunk 2 above)
Comment (calebj) [reply extensions/http-curl/protocols/RESTProtocol.cpp]: The linter wants a constant ref passed, but it doesn't compile due to rapidjson's semantics.

File: extensions/http-curl/protocols/RESTReceiver.cpp (hunk 6 above)
Comment (calebj) [reply extensions/http-curl/protocols/RESTReceiver.cpp]: All of the tests passed, and I made sure the JSON payloads generated by the exchanges were identical. But no, I haven't tested it in a production setting.

**Comment by achristianson** [comment]
Looks like a test failure in the travis xcode:

```
The following tests FAILED:
	 38 - CompressContentTests (Failed)
```

**Comment by achristianson** [comment]
s2s integration tests for this are passing (after a few docker int test fixes --see other PRs).

**Comment by calebj** [comment]
@achristianson something in the master branch is failing. Looks like something in the commit before the latest one, 9b9c3354330d525cfdd50f656ae42fc3da80764c, broke it.

Regarding the LICENSE, there are a handful of them. IANAL, so can someone look over [rapidjson's license.txt](https://github.com/Tencent/rapidjson/blob/master/license.txt) and see what might have to be left out, if anything?

File: extensions/http-curl/protocols/RESTProtocol.cpp (hunk 3 above)
Comment (achristianson) [reply extensions/http-curl/protocols/RESTProtocol.cpp]: IMO we should only use macros where really necessary, so the code can be clearer and easier to read.

**Comment by achristianson** [comment]
@phrocker There are some benchmarks over here: https://github.com/miloyip/nativejson-benchmark

**Comment by achristianson** [comment]
@calebj with regard to the license, I think @phrocker is just referring to adding RapidJSON's license text to our LICENSE file. If you look in there, for every project we include in our thirdparty/ dir, we mention that we did that and what the license of that project is.

RapidJSON is MIT license which is fully compatible. A good reference is https://www.apache.org/legal/resolved.html

So I think we're good if you copy their LICENSE (https://github.com/Tencent/rapidjson/blob/master/license.txt) into our LICENSE using the same format as the others.

**Comment by calebj** [comment]
I know that it's MIT licensed, but their license.txt comes with others tacked on for subcomponents. If you think they're all compatible, I'll copy all of them in.

**Comment by achristianson** [comment]
@calebj will check it out.

File: libminifi/src/RemoteProcessorGroupPort.cpp (hunk 9 above)
Comment (calebj) [reply libminifi/src/RemoteProcessorGroupPort.cpp]: All I did here was translate the jsoncpp code to rapidjson. I didn't test it.

File: extensions/libarchive/UnfocusArchiveEntry.cpp (hunk 8 above)
Comment (calebj) [reply extensions/libarchive/UnfocusArchiveEntry.cpp]: Done in 00282396e2b2fa3f0f450a7916d692b52c749a1d

**Comment by achristianson** [comment]
We should probably make a decision on this soon because this type of change can get out of date fairly easily and JSON touches a lot of things in the codebase.

**Comment by phrocker** [comment]
I'm currently not a +1 on this PR.

**Comment by achristianson** [comment]
Suggest if we are going to merge it, a good time would be the very start of 0.5.0 so we have plenty of time to test and fix any issues that crop up by the time the next release comes around.

**Comment by phrocker** [comment]
@calebj HTTP Site to Site and C2 don't seem to work. Do you need help in testing these?

**Comment by calebj** [comment]
@phrocker Honestly, I just reimplemented the JSON handling in RapidJSON as well as I could, and fixed the things that broke the build and tests. Help would be greatly appreciated.

**Comment by apiri** [comment]
Hey @calebj @phrocker,

Looks like we have a bit of effort to wrap this up remaining so would like to slide this to 0.5.0.  Would like for 0.4.0 to be available so we can provide the appropriate tooling to interact with the RPG changes in NiFi 1.5.0.

**Comment by phrocker** [comment]
@calebj I've rebased this PR so that it's up to date with master and have addressed some issues. I'll try to submit a PR against PR shortly after I finish some final testing.

**Comment by phrocker** [comment]
@calebj I submitted a PR, though it has the full history. Did you ever create a benchmark. While @achristianson linked the benchmark's on github, I've seen slowdowns of 1-2% when parsing our payloads since they are significantly smaller.

**Comment by calebj** [comment]
Squashed.

@phrocker No, I didn't create a benchmark. The slowdown is surprising, though. I can't prove it, but I think larger production payloads will show improvement.
